src/query/storage/promremote/storage.go

```diff
@@ -51,6 +51,56 @@ var errorReadingBody = []byte("error reading body")
 
 var errNoEndpoints = errors.New("write did not match any of known endpoints")
 
+// WriteQueue This is used by a single thread go routine
+type WriteQueue struct {
+	t        tenant
+	capacity int
+	queries  []*storage.WriteQuery
+}
+
+func NewWriteQueue(t tenant, capacity int) *WriteQueue {
+	return &WriteQueue{
+		t:        t,
+		capacity: capacity,
+		queries:  make([]*storage.WriteQuery, 0, capacity),
+	}
+}
+
+func (wq *WriteQueue) pop() []*storage.WriteQuery {
+	res := wq.queries
+	wq.queries = make([]*storage.WriteQuery, 0, wq.capacity)
+	return res
+}
+
+func (wq *WriteQueue) Add(query *storage.WriteQuery) ([]*storage.WriteQuery, error) {
+	if len(wq.queries) >= wq.capacity {
+		return wq.pop(), errors.New(wq.t.name + " write queue is Full")
+	}
+	wq.queries = append(wq.queries, query)
+	return nil, nil
+}
+
+func (wq *WriteQueue) Flush(ctx context.Context, p *promStorage, force bool) {
+	size := len(wq.queries)
+	if size < 10 {
+		if size == 0 || !force {
+			p.logger.Warn("don't do tick flush for small batch with less than 10 samples, "+
+				"samples will be evently written to remote storage as it accumulates.",
+				zap.String("tenant", wq.t.name), zap.Int("size", size))
+			return
```

Comment:
It should return an error to inform the call site the writes are not really flushed.